Expand qq docs re esc sequences

codesections opened the pull request:
This expands the description of interpolation with qq, especially with regards to backslash escape sequences, which it now lists in a table.  It also separates the discussion of variable, closure, and backslash-escape interpolation and emphasizes the connection to the equivalent quoting adverb(s). 

Finally, this also fixes a link on the syntax page that incorrectly implied that backslash escape sequences work in `q` rather than `qq`.

JJ commented on doc/Language/quoting.pod6:
```diff
@@ -174,38 +174,26 @@ My favorite color is blue!
-=begin code :allow<B> :preamble<my $color = 'blue'>
-say B<">The B<\>$color variable contains the value '$color'B<">;
-=end code
+=head3 Interpolating variables
```
I'm not usually in favor of deletion, unless it's patently wrong and it's addressing some issue. That text is usually there for a reason, and it might be a subtle reason, but reasonable nonetheless.

codesections replied:
This code wasn't deleted, just moved.  In this commit, it's at lines 303–315.

JJ replied:
Fair enough. This is a subsection of a section that's called `Interpolation: qq`. Subsections create index entries, so you might want to flesh this out to something that more clearly shows the context or can be found more easily.

JJ reviewed (commented):
While the PR subject says "expand", you're actually changing many paragraphs, eliminating examples (or moving them around). Expansion is good, substitution I'd appreciate first an issue that says why it fails. Rewriting for the sake of rewriting might eliminate some stuff that's there for a good reason.

codesections commented:
That's a fair point and I probably should have opened an issue first.  (I started this PR intending to just add a table of allowed backslash escape sequences, but then couldn't find a good place to put it and ended up moving more text around than I originally meant to.)

That said, I don't believe I deleted anything – I just moved a few things around to create some subsections within the `Interpolation: qq` section.

Let me know if you'd like me to open an issue, though; I'd be happy to do so.

JJ commented:
> That's a fair point and I probably should have opened an issue first. (I started this PR intending to just add a table of allowed backslash escape sequences, but then couldn't find a good place to put it and ended up moving more text around than I originally meant to.)
> 
> That said, I don't believe I deleted anything – I just moved a few things around to create some subsections within the `Interpolation: qq` section.
> 
> Let me know if you'd like me to open an issue, though; I'd be happy to do so.

Fair enough. Let me check out what you've done. I'd also appreciate, if possible, that you made this a bit clearer in the PR subject.

JJ commented on doc/Language/quoting.pod6:
```diff
@@ -174,38 +174,26 @@ My favorite color is blue!
-written within the string so that the content of the variable is inserted into
-the string. It is also possible to escape variables within a C<qq>-quoted
-string:
+interpolation of backslash escape sequences (like C<q:backslash>), all sigiled
```
nit: just scratch "all"

JJ commented on doc/Language/quoting.pod6:
```diff
@@ -174,38 +174,26 @@ My favorite color is blue!
-string:
+interpolation of backslash escape sequences (like C<q:backslash>), all sigiled
+variables (like C<q:scalar:array:hash:function>), and any code inside C<{...}>
+(like C<q:closure>).
```
I was a bit confused here, but then I understood that what you want to say is that `qq` works in the same way as q with a series of adverbs. It might be clear withing the whole context of the page, but either a link or a fleshed out clarification might be clearer.

JJ commented on doc/Language/quoting.pod6:
```diff
@@ -174,38 +174,26 @@ My favorite color is blue!
-=begin code :lang<text>
-The $color variable contains the value 'blue'
-=end code
+Inside a C<qq>-quoted string, you can use variables with a sigil to trigger
```
This is the second time you mention variables `with a sigil`; however, if they don't have a sigil, how can they be identified as variables? Also, arguably you're talking about identifiers, not variables. Something with a sigil is clearly an identifier; no sigil, well, it could be, but not always.

JJ commented on doc/Language/quoting.pod6:
```diff
@@ -174,38 +174,26 @@ My favorite color is blue!
-=end code
+Inside a C<qq>-quoted string, you can use variables with a sigil to trigger
+interpolation of the variable's value.  Variables with the C<$> sigil are
+interpolated whenever the occur (unless escaped); that's why, in the example
```
typo: they

JJ commented on doc/Language/quoting.pod6:
```diff
@@ -174,38 +174,26 @@ My favorite color is blue!
-This way, when you write C<"documentation@raku.org">, you aren't
-interpolating the C<@raku> variable. If that's what you want to do, append
-a C<[]> to the variable name:
+To interpolate an Array (or other L<Positional|/type/Positional> variable),
```
Probably just talk about anything with the `@` sigil instead. I really have no idea why this happens, BTW

JJ commented on doc/Language/quoting.pod6:
```diff
@@ -174,38 +174,26 @@ My favorite color is blue!
-interpolating the C<@raku> variable. If that's what you want to do, append
-a C<[]> to the variable name:
+To interpolate an Array (or other L<Positional|/type/Positional> variable),
+append a C<[]> to the variable name:
```
Apparently, same goes for `%`

```
> my %h = :a, :b, :c
{a => True, b => True, c => True}
> say "%h"
%h
> say "%h[]"
a	True
b	True
c	True
> 
```

JJ commented on doc/Language/quoting.pod6:
```diff
@@ -216,9 +204,9 @@ say "@neighborsB<[]> and I try our best to coexist peacefully."
-code will work:
+Alternatively, rather than using C<[]>, you can interpolate the Array by
+following it with a method call with parentheses after the method name. Thus the
+following code will work:
```
All this seems strangely ad hoc and does not seem to match the "strangely consistent" nature of Raku. There must be some underlying simple rule we're not getting

JJ commented on doc/Language/quoting.pod6:
```diff
@@ -230,20 +218,63 @@ Felix, Danielle, Lucinda and I try our best to coexist peacefully.
 
-Postcircumfix operators and therefore L<subscripts|/language/subscripts> are
-interpolated as well.
+To interpolate a Hash (or other L<Associative|/type/Associative> variable), use
```
As shown above, it also works with []

JJ commented on doc/Language/quoting.pod6:
```diff
@@ -230,20 +218,63 @@ Felix, Danielle, Lucinda and I try our best to coexist peacefully.
+
+=begin table :allow<L>
+Sequence    Hex Value       Character
+\0          \x0000          L<Nul|https://util.unicode.org/UnicodeJsps/character.jsp?a=0000&B1=Show>
```
I've usually seen this in ALL CAPS, right?

JJ commented on doc/Language/syntax.pod6:
```diff
@@ -568,8 +568,8 @@ String literals are surrounded by quotes:
 See L<quoting|/language/quoting> for many more options, including
-L<the escaping quoting C<q>|/language/quoting#Escaping:_q>. Raku uses
-the standard escape characters in literals: C<\a \b \t \n \f \r \e>,
+L<interpolation quoting C<q>|/language/quoting#Interpolation:_qql>. Raku uses
```
Thanks for taking care of links.

JJ reviewed (approved):
Except for typos, good job. Still baffled at the interpolation rules, though. Will postcircumfix operators trigger it?

JJ commented:
I have been testing with defined postcircumfix operators and it's not their presence what triggers the interpolation; it's apparently either (), [] or <>. So, no idea.

TimToady commented:
Look at the use and definition of the bracket_ending rule in Grammar.nqp.

Larry

On Sun, Jan 16, 2022 at 10:51 AM Juan Julián Merelo Guervós <
***@***.***> wrote:

> I have been testing with defined postcircumfix operators and it's not
> their presence what triggers the interpolation; it's apparently either (),
> [] or <>. So, no idea.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Raku/doc/pull/4015#issuecomment-1013931045>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABHSYWKXI2DHT6NJO5LYNDUWMHRPANCNFSM5L75D6JA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
